main/src/ui/util/helper.vala

```diff
@@ -144,59 +144,87 @@ public static bool is_24h_format() {
     return settings_format == "24h" || p_format == " ";
 }
 
-public static string parse_add_markup(string s_, string? highlight_word, bool parse_links, bool parse_text_markup, bool already_escaped_ = false) {
-    string s = s_;
-    bool already_escaped = already_escaped_;
+public static string parse_add_markup(string s_, string? highlight_word, bool parse_links, bool parse_text_markup) {
+    Gee.ArrayList<int> start_url_pos = new Gee.ArrayList<int>();
+    Gee.ArrayList<int> end_url_pos = new Gee.ArrayList<int>();
+    string s = Markup.escape_text(s_);
 
     if (parse_links) {
-        Regex url_regex = new Regex("""(?i)\b((?:[a-z][\w-]+:(?:\/{1,3}|[a-z0-9%])|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}\/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'".,<>?«»“”‘’]))""");
-        MatchInfo match_info;
-        url_regex.match(s.down(), 0, out match_info);
-        if (match_info.matches())  {
+        Regex url_regex = new Regex("""(?i)((https?|ftp|magnet|tel)+):(?:\/\/(?:((?:[a-z0-9-._~!$&'()*+,;=:]|%[0-9A-F]{2})*)@)?((?:[a-z0-9-._~!$'()*+,;=]|%[0-9A-F]{2})*)(?::(\d*))?(\/(?:[a-z0-9-._~!$&'()+,;=:@/]|%[0-9A-F]{2})*)?|(\/?(?:[a-z0-9-._~!$&'()*+,;=:@]|%[0-9A-F]{2})+(?:[a-z0-9-._~!$&'()*+,;=:@/]|%[0-9A-F]{2})*)?)(?:\?((?:[a-z0-9-._~!$&'()*+,;=:/?@]|%[0-9A-F]{2})*))?(?:#((?:[a-z0-9-._~!$&'()*+,;=:/?@]|%[0-9A-F]{2})*))?""");
```

Comment:
Can't this be generalized? I mean there are tons of other URI schemes.. You forgot `mailto:` f.e. Another one I personally use a lot is `bitcoin:`. Does the RFC not provide a general one? Found one: https://tools.ietf.org/html/rfc3986#appendix-B.

Reply:
@stevenroose I tried not to match the whole world but it's as simple as adding it at the begining (as you noticed).
It's also hard to generalize it while maintaining a decent match to actual URLs. (the IETF one doesn't work)